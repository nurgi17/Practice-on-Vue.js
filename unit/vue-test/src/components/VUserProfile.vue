<template>
  <div class="user-profile">
    <img class="user-profile__avatar" :src="user.avatar_url" alt="">
    <div class="user-profile__name">
      {{ user.name }}
    </div>
    <div class="user-profile__bio">
      {{ user.bio }}
    </div>
  </div>
</template>

<script>
export default {
  props: {
    user: {
      type: Object,
      required: true,
      default: () => ({})
    }
  }
}
</script>

<style scoped>
.user-profile {
  border-top: 1px solid #ccc;
  padding-top: 20px;
  margin-top: 20px;
  display: flex;
  flex-direction: column;
  align-items: center;
}
.user-profile__avatar {
  border-radius: 50%;
  width: 150px;
  height: 150px;
}
.user-profile__name {
  margin-top: 14px;
  font-size: 24px;
  font-family: 'Bungee', cursive;
  text-transform: uppercase;
}
.user-profile__bio {
  margin-top: 14px;
  font-family: 'Open Sans', sans-serif;
}
</style>
