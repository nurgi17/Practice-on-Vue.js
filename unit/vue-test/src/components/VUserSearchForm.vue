<template>
  <form @submit.prevent="$emit('submitted', username)">
    <el-input
      placeholder="Enter github username"
      v-model="username"
      class="search-form__input"
    >
      <el-button
        class="search-form__button"
        native-type="submit"
        slot="append"
        icon="el-icon-search"
      ></el-button>
    </el-input>
  </form>
</template>

<script>
export default {
  data () {
    return {
      username: ''
    }
  }
}
</script>
