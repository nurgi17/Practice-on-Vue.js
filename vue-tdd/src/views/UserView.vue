<template>
  <div>
    Hello nurgi17
  </div>
</template>

<script>
export default {
  name: 'UserView'
}
</script>
