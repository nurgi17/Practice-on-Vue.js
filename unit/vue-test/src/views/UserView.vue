<template>
  <div>
    <VUserSearchForm @submitted="searchUser"/>
    <VUserProfile :user="user" style="margin-top: 20px;"/>
  </div>
</template>

<script>
import { mapState } from 'vuex'
import VUserProfile from '@/components/VUserProfile.vue'
import VUserSearchForm from '@/components/VUserSearchForm.vue'

export default {
  components: {
    VUserProfile,
    VUserSearchForm
  },
  // data () {
  //   return {
  //     user: {
  //       name: ''
  //     }
  //   }
  // },
  computed: {
    ...mapState({
      user: 'user'
    })
  },
  methods: {
    searchUser (username) {
      this.$store.dispatch('SEARCH_USER', { username })
    }
  }
}
</script>
