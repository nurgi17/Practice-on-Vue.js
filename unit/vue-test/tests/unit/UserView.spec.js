jest.mock('@/store/actions')
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import UserView from '@/views/UserView'
import VUserProfile from '@/components/VUserProfile.vue'
import VUserSearchForm from '@/components/VUserSearchForm.vue'
import initialState from '@/store/state'
import actions from '@/store/actions'
import userFixture from './fixtures/user'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('UserView', () => {
  let state
  const build = () => {
    const wrapper = shallowMount(UserView, {
      localVue,
      store: new Vuex.Store({ 
        state,
        actions,
      })
      // data () {
      //   return {
      //     user: {}
      //   }
      // },
    })

    return {
      wrapper,
      userSearchForm: () => wrapper.findComponent(VUserSearchForm),
      userProfile: () => wrapper.findComponent(VUserProfile)
    }
  }

  beforeEach(() => {
    jest.resetAllMocks()
    state = { ...initialState }
  })

  it('renders component', () => {
    // arrange
    const { wrapper } = build()

    // assert
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('renders main child components', () => {
    // arrange
    const { userSearchForm, userProfile } = build()

    // assert
    expect(userSearchForm().exists()).toBe(true)
    expect(userProfile().exists()).toBe(true)
  })
  it('passes binded user prop to user profile component', () => {
    // arrange
    // const { wrapper, userProfile } = build()
    // wrapper.setData({
    //   user: {
    //     name: 'Nurgissa'
    //   }
    // })
    state.user = userFixture
    const { userProfile } = build()

    // assert
    expect(userProfile().vm.user).toBe(state.user)
  })
  it('searches for a user when received "submitted"', () => {
    // arrange
    const expectedUser = 'nurgi17'
    const { userSearchForm } = build()

    // act
    userSearchForm().vm.$emit('submitted', expectedUser)

    // assert
    expect(actions.SEARCH_USER).toHaveBeenCalled()
    expect(actions.SEARCH_USER.mock.calls[0][1]).toEqual({ username: expectedUser })
  })
})