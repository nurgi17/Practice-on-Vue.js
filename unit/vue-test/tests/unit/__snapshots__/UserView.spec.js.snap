// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserView renders component 1`] = `
<div>
  <vusersearchform-stub></vusersearchform-stub>
  <vuserprofile-stub user="[object Object]" style="margin-top: 20px;"></vuserprofile-stub>
</div>
`;
